### ISSUE TITLE
Wrong input from multipart/form-data

### ISSUE THREAD

**Issue by thetutlage** [adonisjs/adonis-middleware#10]
_From @andrasf on March 31, 2017 13:23_

Hi,

If I send input with ampersand sign from a form with files: true, I only get the part before the & sign.
So as a simple test 1&2 becomes 1.
My problem is I need to submit encoded html values from an editor, which only get received until the first &amp; entity.

Thanks,
András

_Copied from original issue: adonisjs/adonis-framework#529_

**Comment by thetutlage** [comment]
Mind sharing the code where you are reading these values?

**Comment by thetutlage** [comment]
_From @andrasf on March 31, 2017 15:15_

I read a list of related datas, but its simply
```
description = request.input(`description-${i}`, '')
console.log(description)
```
If my input was `1&2`, `1` is logged to the console.

In Chrome Dev Tools is see
```
------WebKitFormBoundarypbXmRTmeTjxIfZrM
Content-Disposition: form-data; name="description-0"

1&2
```

**Comment by thetutlage** [comment]
_From @andrasf on April 1, 2017 7:48_

I created a new empty project.

routes.js:
```
'use strict'

const Route = use('Route')

Route.get('/', 'TestController.index')
Route.post('/', 'TestController.print')
```

TestController:
```
'use strict'

class TestController {
  * index (request, response) {
    yield response.sendView('welcome')
  }

  * print (request,response) {
    response.send(request.input('stuff'))
  }
}

module.exports = TestController
```

welcome.njk:
```
{% extends 'master' %}

{% block content %}
  {{ form.open({ action: 'TestController.print', files: true }) }}
    {{ csrfField }}
    <input name="stuff" />
    <input type="submit" />
  {{ form.close() }}
{% endblock %}
```

It acts the same way. Works fine without `files: true`.

**Comment by thetutlage** [comment]
Identified bug, working on the fix

**Comment by thetutlage** [comment]
Install `adonis-middleware` from github and lemme if all works fine for you?

**Comment by andrasf** [comment]
I've checked and it works.

Also checked it with file uploads, I've got one notice: when uploading multiple files but not leaving some file inputs empty mixes the order.
If there are 3 inputs and 3 files (1, 2, 3) then uploading `1, 2, 3` results in order `1, 2, 3`.
Uploading `1, empty, empty` results in `empty, empty, 1`.
Uploading `1, empty, 3` results in `empty, 1, 3`.

Looks like empty ones always come first.

**Comment by andrasf** [comment]
@thetutlage what do you think?

**Comment by thetutlage** [comment]
Ohhh can u share a example form?

**Comment by andrasf** [comment]
I pushed an example app to https://github.com/andrasf/amw-test
I'll try to explain...

Input:
![0000-input](https://cloud.githubusercontent.com/assets/288001/24612269/0c6af81e-1885-11e7-96c3-9e736d5320da.png)

Output:
![0000-output](https://cloud.githubusercontent.com/assets/288001/24612272/0eeaf35a-1885-11e7-8f5c-520a5d8d0e98.png)

If you visit url localhost:3333 and select only first file on the first form, so you send 1-first, empty, empty, then you should get the output  empty, empty, 1-first, the order changes. (There are 3 files in the root folder: 1-first, 2-second and 3-third).

Other question:
I used field names without square bracket on the first form.

If field names are `texts` and `files` I can access them as arrays by `request.input('texts')` and `request.file('files')`
If field names are `texts[]` and `files[]` I can access them as arrays by `request.input('texts')` and `request.file('files[]')`

What seems not to work is that field names are `texts[]` and `files[]` and access them by `request.input('texts')` and `request.file('files')`. That's what you can see in second form and `print2` method.

**Comment by thetutlage** [comment]
I believe browser prioritizes the one's with content when you do not define explicit indexes. This is how the payload of chrome looks like

```
------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="_csrf"

vDH4EIPc-ljeOxVKJ9Ct8YNVYFrtRLOqkXkM
------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="texts"

1&2
------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="texts"

2&2
------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="texts"

3&2
------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="files[]"; filename="_3-1.jpg"
Content-Type: image/jpeg


------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="files[]"; filename=""
Content-Type: application/octet-stream


------WebKitFormBoundaryAcTIRTu2CtggO3lR
Content-Disposition: form-data; name="files[]"; filename=""
Content-Type: application/octet-stream


------WebKitFormBoundaryAcTIRTu2CtggO3lR--
```

The best way always is to have `[0]`, `[1]` indexes on array fields

**Comment by thetutlage** [comment]
For sure there is race condition and that is inside the `formidable` code base. Adding explicit indexes to the array files will fix this issue.

```html
<input name="files[0]" type="file" /><br />
<input name="files[1]" type="file" /><br />
<input name="files[2]" type="file" /><br />
```

You can create an issue with https://github.com/felixge/node-formidable if you want. I am stopping myself, since I have plans not to use formidable in new major release, since it has bunch of edge case.

**Comment by andrasf** [comment]
I see.

Does this mean multiple inputs with `name="texts"` (without square brackets) being accessible as an array with `request.input('texts')` work fine this way? This part is blurry for me, shouldn't they be `name="texts[]"`?

Otherwise the original issue is solved.

**Comment by andrasf** [comment]
Sorry, it looks like it works both ways now.

**Comment by andrasf** [comment]
That's nice, thanks for the fix.

**Comment by thetutlage** [comment]
Ideally the best way to have array like fields as `texts[]`. Now the `index` number (texts[0]) is not required.

But with formidable, they mess up the order when you do not define the index and hence we need it.

**Comment by andrasf** [comment]
I mean multiple `name="texts"` fields are also accessible as array (without square bracket).

**Comment by thetutlage** [comment]
Okay. I will release this now.

**Comment by thetutlage** [comment]
Released as 1.0.13
